rust/connlib/tunnel/src/lib.rs

```diff
@@ -139,11 +139,23 @@ impl ClientTunnel {
                     continue;
                 }
                 Poll::Ready(io::Input::Device(packet)) => {
+                    let src = packet.source();
+                    let dst = packet.destination();
+
                     let Some(transmit) = self.role_state.encapsulate(packet, Instant::now()) else {
                         continue;
                     };
 
-                    self.io.send_network(transmit)?;
+                    match self.io.send_network(transmit) {
+                        Ok(()) => {}
+                        Err(e) if is_network_unreachable(&e) => {
+                            let response =
+                                ip_packet::make::icmp_network_unreachable(dst, src).unwrap();
+
+                            self.io.send_device(response)?;
+                        }
+                        Err(e) => return Poll::Ready(Err(e)),
+                    }
```

Comment:
This is the same codepath that we also use for sending encrypted WG packets. We should probably add a safe-guard here to ensure we only do this for forwarded DNS queries in order to not disrupt any data plane connections during roaming.

Reply:
I think the way I want to solve this is by passing the `io::Error` to the state instead, together with the destination of the datagram.

We can then internally look up whether or not that was a forwarded DNS query and if it is, we can construct an ICMP error. That should be more robust and most importantly, can be tested as part of tunnel test.